sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -177,6 +177,37 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
     if (allInputBoxesIds.size != allInputBoxesIds.toSet.size) {
       throw new IllegalArgumentException(s"Block ${mod.id} contains duplicated input boxes to open")
     }
+
+    checkWithdrawalBoxesAllowed(mod)
+  }
+
+  private def checkWithdrawalBoxesAllowed(mod: SidechainBlock): Unit = {
+    val alreadyMinedWBs = getAlreadyMinedWithdrawalRequestBoxesInCurrentEpoch
+    val mainchainBlockReferenceInBlock = mod.mainchainBlockReferencesData.size
+    val allowedWBs = getAllowedWithdrawalRequestBoxes(mainchainBlockReferenceInBlock)
+    var blockWBs = alreadyMinedWBs
+    mod.transactions.foreach(tx => {
+      blockWBs += tx.newBoxes().asScala.count(box => box.isInstanceOf[WithdrawalRequestBox])
+      if (blockWBs > allowedWBs) {
+        throw new IllegalStateException(s"Exceeded the maximum number of WithdrawalBoxes allowed!")
+      }
+    })
+  }
+
+  def getOrElseWithdrawalEpochInfo(): WithdrawalEpochInfo = {
```

Comment:
check if can keep private any of these new methods if not called from outside

Reply:
Done